src/mon/MDSMonitor.cc

```diff
@@ -1641,7 +1641,9 @@ int MDSMonitor::load_metadata(map<mds_gid_t, Metadata>& m)
   bufferlist bl;
   int r = mon->store->get(MDS_METADATA_PREFIX, "last_metadata", bl);
   if (r) {
-    dout(1) << "Unable to load 'last_metadata'" << dendl;
+    if (r != -ENOENT) {
```

Comment:
this doesn't sound right. The underlying store is only propagating one error via the return value, and that is ENOENT. What you are doing is silently creating a dead branch that will *always* suppress the error message.

I'm not sure what problem you are trying to address, but this doesn't seem the right way to do it.

Reply:
thanks for your review @jecluis   
logging this message when quering cyclicly(ceph mds versions) seems useless and wastes space. we'd better silence it at least in normal use case.

i can think of two ways to optimize this as only ENOENT exists,  the first one is just return,(https://github.com/ceph/ceph/blob/0ccdc799384801faa728d5429e82b522ee3b618b/src/mon/MgrMonitor.cc#L1045); the second one is higher log lever, maybe say ldout(10) ?

which one do you prefer? or any other ideas ?

Reply:
@jecluis ping

Reply:
@tchaikov as it seems jecluis busy with other things., please take a review if you have time.

Reply:
will do the day after tomorrow.

Reply:
@shun-s sorry, was on vacation for the past couple of weeks.

I may be understanding this code path incorrectly, but it seems that 'last_metadata' should exist as long as at least one mds has booted at some point. It doesn't seem to be erased at any point after being created, and (at most) entries are removed from the metadata itself.

So I'm a bit curious how you are getting to a point where it makes sense to suppress these messages. You mentioned 'ceph mds version', so I'm inclined to assume that you are seeing this in a cluster that may or may not be using cephfs, and may or may not have a populated mdsmap, but, regardless, has never had a running mds. Is that correct?

If my assumption is correct, and you never had an mds running, I think the way to suppress this warning is something that should be taken care somewhere else (maybe the caller, checking if we have a populated mdsmap?). If my assumption is incorrect, I think there may be some other problem, because I /think/ that if you have at least one mds running then you *need* to have that key in the store.

Reply:
yeah, your assumption is correct, but
1.  with or without this message doesn't make much sense ;
2.  to detect if there are mds services,  one way is to config the manager daemon (as you said) after cluster deploed without mds for a period, which requires restart services and interrupt service; the other way is to periodically use `ceph mds xx` command, which is more friendly and easy.

so i suggest to suppress these messages.

Reply:
I don't think we can, in good conscious, suppress these messages. The message is there to make it clear that the key does not exist, and I think it is very worthwhile for those cases where it is supposed to be there.

I think the best solution is to check whether there are mds in the mdsmap, and only call `count_metadata()` if so, for all the callers of `count_metadata()`. Or, maybe simpler, do that in `count_metadata()` itself. I think this will require adding a getter method in FSMap though (something like `has_daemons()`), because I don't think we currently expose this easily without calling something like `FSMap::get_mds_info()` and checking if it's empty.

Reply:
!_!,  the best solution you proposed is not ready now, and i agree with the opinion that these messages are worthwhile. 
     could we just increase the dout level, say from dout(1) -> dout(5) ?
because on one hand, these message are worthwhile but not urgent,  on the other hand,  these messages will be slienced at usual.
     anyway, thanks for your detailed explaination and guidance.

[词典] | guidance
-- | --

Reply:
@jecluis  ping

Reply:
@jecluis repush done.